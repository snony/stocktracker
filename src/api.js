const host = 'https://api.iextrading.com/1.0/'

const fetchAndUnpack = url => fetch(url).then(data => data.json())

export const getHistory = (symbol, dateFilter = 'ytd', priceFilter = 'close') => {
  const url = `${host}stock/${symbol}/chart/${dateFilter}?filter=date,${priceFilter}`
  return fetchAndUnpack(url)
}
export const getRefData = () => {
  const url = `${host}ref-data/symbols?filter=name,symbol`
  return fetchAndUnpack(url)
}

export const getNews = symbol => {
  const url = `${host}stock/${symbol}/news/last/5`
  return fetchAndUnpack(url)
}

export const getOverview = symbol => {
  const url = `${host}stock/${symbol}/company`
  return fetchAndUnpack(url)
}

export const getPeers = symbol => {
  const url = `${host}stock/${symbol}/peers`
  return fetchAndUnpack(url)
}

export const getStats = symbol => {
  const url = `${host}stock/${symbol}/stats`
  return fetchAndUnpack(url)
}

export const getPrevious = symbol => {
  const url = `${host}stock/${symbol}/previous`
  return fetchAndUnpack(url)
}
export const getOhlc = symbol => {
  const url = `${host}stock/${symbol}/ohlc`
  return fetchAndUnpack(url)
}
export const getQuote = symbol => {
  const url = `${host}stock/${symbol}/quote`
  return fetchAndUnpack(url)
}

const keyStatsServices = [getStats, getPrevious, getOhlc, getQuote]

const extractData = ([stats, previous, ohlc, quote]) => ({
  earningsPerShare: stats.latestEPS,
  dividendYield: stats.dividendYield,
  previousClose: previous.close,
  open: ohlc.open.price,
  close: ohlc.close.price,
  dayHigh: ohlc.high,
  dayLow: ohlc.low,
  volume: quote.latestVolume,
  avgTotalVolume: quote.avgTotalVolume,
  peRatio: quote.peRatio,
  marketCap: quote.marketCap,
  week52High: quote.week52High,
  week52Low: quote.week52Low
})

export const getKeyStats = symbol => {
  return Promise.all(keyStatsServices.map(service => service(symbol))).then(extractData)
}

const companyInfoServices = [getHistory, getNews, getOverview, getPeers, getKeyStats]
const extractCompanyInfo = ([history, news, overview, peers, keystats]) => ({
  history,
  news,
  overview,
  peers,
  keystats
})

export const getCompanyInfo = symbol => {
  return Promise.all(companyInfoServices.map(service => service(symbol))).then(extractCompanyInfo)
}
