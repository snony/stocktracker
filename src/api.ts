const host = 'https://api.iextrading.com/1.0'

const fetchAndUnpack = async <T>(url: string): Promise<T> => {
  const data = await fetch(url)
  return await data.json()
}

export const getCompanySymbols = async () => {
  const url = `${host}/ref-data/symbols?filter=name,symbol`
  return await fetchAndUnpack(url)
}

export const getHistory = async (
  symbol: string,
  dateFilter: string = 'ytd',
  priceFilter: string = 'close'
) => {
  const url = `${host}/stock/${symbol}/chart/${dateFilter}?filter=date,${priceFilter}`
  return await fetchAndUnpack(url)
}

export const getNews = async (symbol: string) => {
  const url = `${host}/stock/${symbol}/news/last/5`
  return await fetchAndUnpack(url)
}

export const getOverview = async (symbol: string) => {
  const url = `${host}/stock/${symbol}/company`
  return await fetchAndUnpack(url)
}

export const getPeers = async (symbol: string) => {
  const url = `${host}/stock/${symbol}/peers`
  return await fetchAndUnpack(url)
}

interface StockData {
  [key: string]: number
}

interface KeyStatsData {
  quote: StockData
  stats: StockData
}

export const extractData = ({ quote, stats }: KeyStatsData) => ({
  earningsPerShare: stats.latestEPS,
  dividendYield: stats.dividendYield,
  previousClose: quote.previousClose,
  open: quote.open,
  dayHigh: quote.high,
  dayLow: quote.low,
  volume: quote.latestVolume,
  avgTotalVolume: quote.avgTotalVolume,
  peRatio: quote.peRatio,
  marketCap: quote.marketCap,
  week52High: quote.week52High,
  week52Low: quote.week52Low
})

export const getKeyStats = async (symbol: string) => {
  const url = `${host}/stock/${symbol}/batch?types=quote,stats`
  const keystats = await fetchAndUnpack<KeyStatsData>(url)
  return extractData(keystats)
}

export default {
  getCompanySymbols,
  getHistory,
  getNews,
  getKeyStats,
  getOverview,
  getPeers
}
