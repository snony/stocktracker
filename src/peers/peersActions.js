export const PEERS_RECEIVED_ACTION = 'PEERS_RECEIVED_ACTION'

const peersReceivedAction = peers => ({ type: PEERS_RECEIVED_ACTION, peers })

export const getPeersData = symbol => (dispatch, _, api) => {
  api.getPeers(symbol).then(peers => {
    return dispatch(peersReceivedAction(peers))
  })
}
