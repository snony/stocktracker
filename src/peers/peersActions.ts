import { Action, ActionCreator } from 'redux'
import { ThunkAction } from 'redux-thunk'

import { API, GlobalState } from '../types'
import { Peers } from './types'

export const PEERS_RECEIVED_ACTION = 'PEERS_RECEIVED_ACTION'

export interface PeersReceivedAction extends Action {
  type: typeof PEERS_RECEIVED_ACTION
  peers: Peers
}

const peersReceivedAction: ActionCreator<PeersReceivedAction> = (peers: string[]) => ({
  type: PEERS_RECEIVED_ACTION,
  peers
})

type ThunkResult<R> = (symbol: string) => ThunkAction<R, GlobalState, API, PeersReceivedAction>

export const getPeersData: ThunkResult<void> = (symbol: string) => (dispatch, _, api) => {
  api.getPeers(symbol).then(peers => dispatch(peersReceivedAction(peers)))
}
