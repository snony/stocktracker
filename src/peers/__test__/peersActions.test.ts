import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { getPeersData, PEERS_RECEIVED_ACTION, PeersReceivedAction, peersReceivedAction } from '../peersActions'
import api from './mockAPI'
import mockPeersData from './mockData'

const mockSymbol = 'aapl'
const middleware = [thunk.withExtraArgument(api)]
const mockStore = configureMockStore(middleware)
const store = mockStore({ peers: mockPeersData.emptyPeers })
const expectedAction: PeersReceivedAction = {
  type: PEERS_RECEIVED_ACTION,
  peers: mockPeersData.examplePeers
}

describe('test for peers action', () => {
  const mockDispatch = store.dispatch

  it('should create new PEERS_RECEIVED_ACTION object', () => {
    const action = peersReceivedAction(mockPeersData.examplePeers)
    expect(action).toEqual(expectedAction)
  })

  it('should dispatch PEERS_RECEIVED_ACTION when the fetch is successful', () => {
    mockDispatch<any>(getPeersData(mockSymbol)).then(() => {
        const mockStoreAction = store.getActions()
        expect(mockStoreAction[0] as object[]).toEqual(expectedAction)
      })
  })
})

describe('test for peers action without mock-redux-store', () => {
  const mockDispatch = jest.fn()
  const result = getPeersData(mockSymbol)
  return result(mockDispatch, null, api).then(() => {
    expect(mockDispatch.mock.calls).toMatchSnapshot()
  })
})
