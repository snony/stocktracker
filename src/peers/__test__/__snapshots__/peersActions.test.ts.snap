// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Array [
    Object {
      "peers": Array [
        "peers1",
        "peers2",
        "peers3",
        "peers4",
      ],
      "type": "PEERS_RECEIVED_ACTION",
    },
  ],
]
`;
